### PR TITLE
fix m3 duration and value bucketing

### PR DESCRIPTION
**Pull request by amoore877**
Signed-off-by: Andrew Moore <andrew.s.moore@uber.com>

**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [X] Documentation updated?

**Affected functionality**
M3 telemetry bucketing

**Description of change**
Existing bucketing defines no buckets, resulting in a single `minus infinity to infinity` bucket for durations and values. This is not useful.
Define actual latency and value bucketing (use defaults).

**Which issue this PR fixes**
N/A

### CODE REVIEW

File: pkg/common/telemetry/m3.go [hunk 1]
```diff
@@ -11,6 +11,68 @@ import (
 	"github.com/uber-go/tally/m3"
 )
 
+var (
+	// buckets with decreasing granularity as latency increases
+	// given these values, it is much easier to read in this format
+	// than to use the iterative `MakeBuckets` functionality from tally
+	durationBuckets = tally.DurationBuckets{
```
Comment (amoore877) [inline pkg/common/telemetry/m3.go]: this is an initial suggestion for latency buckets, open to suggestions on different bucket cut offs

File: pkg/common/telemetry/m3.go [hunk 2]
```diff
@@ -11,6 +11,68 @@ import (
...
+		2500 * time.Millisecond,
+		3000 * time.Millisecond,
+		4000 * time.Millisecond,
+		5000 * time.Millisecond,
+		10000 * time.Millisecond,
+	}
+
+	// buckets for orders of magnitude of values, up to 1M
+	// given nature of SPIRE, we do not expect negative values
+	valueBuckets = append(tally.ValueBuckets{0}, tally.MustMakeExponentialValueBuckets(1, 10, 6)...)
```
Comment (amoore877) [inline pkg/common/telemetry/m3.go]: initial suggestion for buckets. open to suggestion
this is applicable only to `AddSample` which is today used only  for:
- `DiscoveredSelectors` on workload attestation
- `ExpiringSVIDs` on agent cache manager

File: pkg/common/telemetry/m3.go (hunk 1 above)
Comment (zmt) [reply pkg/common/telemetry/m3.go]: It looks like a decent guess on the face of it. Varying bucket size can get really complicated.

File: pkg/common/telemetry/m3.go [hunk 3]
```diff
@@ -11,6 +11,67 @@ import (
...
+		time.Second,
+		1500 * time.Millisecond,
+		2 * time.Second,
+		3 * time.Second,
+		4 * time.Second,
+		5 * time.Second,
+		10 * time.Second,
+	}
+
+	// buckets for orders of magnitude of values, up to 1M
```
Comment (jkanywhere) [inline pkg/common/telemetry/m3.go]: what is 1M in this context? Minute? Meter? Megabyte?

File: pkg/common/telemetry/m3.go (hunk 2 above)
Comment (jkanywhere) [reply pkg/common/telemetry/m3.go]: `valueBuckets` is a vague sounding name. What is it intended to be used for? `exponentialValueBuckets` is slightly better.

File: pkg/common/telemetry/m3.go [hunk 4]
```diff
@@ -11,6 +11,67 @@ import (
...
+	// given nature of SPIRE, we do not expect negative values
```
Comment (jkanywhere) [inline pkg/common/telemetry/m3.go]: Does it makes sense to add a `-Infinity to 0` bucket so these unexpected cases can be detected?

**Comment by jkanywhere** [comment]
I also envision a duration bucket for SVID TTL, something with granularity as low as minutes and as high as days.
That way we can log and detect when agents or workloads are not getting SVID refresh.

Maybe something like
```
0,
5 * time.Minute,
30 * time.Minute,
1 * time.Hour,
2 * time.Hour,
4 * time.Hour,
6 * time.Hour,
8 *time.Hour,
12 * time.Hour,
1 * time.Day,
2 * time.Day,
5 * time.Day,
```
Might even make sense to have some negative buckets to detect expired certs. You may want it to be symmetric with the positive buckets, or have fewer larger buckets with lower granularity.
```
-Infinity,
-5 * time.Day,
-1 * time.Day,
-6 * time.Hour,
-1 * time.Hour,
-5 * time.Minute, 
```

Does it make sense to add those in the same diff?

**Comment by jkanywhere** [comment]
> I also envision a duration bucket for SVID TTL
> ```
...
> 12 * time.Hour,
> 1 * time.Day,
> 2 * time.Day,
> 5 * time.Day,
> ```

Assuming we log whenever we receive and SVID, and periodically thereafter.
If expected SVID validity period is 1 day, it probably makes sense to add some additional more granular buckets around that time range. Also maybe around the one-half TTL range.
If those values are configurable, maybe the buckets would change too.
```
...
4 * time.Hour,
5 * time.Hour,
6 * time.Hour,
7 *time.Hour,
8 *time.Hour,
12 * time.Hour,
23 * time.Hour,
24 * time.Hour,
25 * time.Hour,
2 * time.Day,
5 * time.Day,
```
We would expect things to stay largely within the 1/2 TTL to 1 TTL range.

File: pkg/common/telemetry/m3.go (hunk 4 above)
Comment (ZymoticB) [reply pkg/common/telemetry/m3.go]: -Infinity to your lowest bucket implicitly exist. As does your highest bucket to infinity

**Comment by ZymoticB** [comment]
There is utility in making these useful in the short term, but the real problem is that the metrics facade prevents the callsite that is recording the duration sample from providing an expectation of the buckets. 

For example, an SVID TTL is a duration, as is the response time of making a request to the datastore plugin. The granularity of buckets in these two situations will vastly differ. Trying to pick one set of buckets to rule them all will likely be problematic in the future.

The preferred usage for tally is to provide your own buckets to  Scope.Histogram and cache that object locally. We'll need to investigate going forward how to best accomplish this without sacrificing the ability to use other metrics backends.

**Comment by amoore877** [comment]
> For example, an SVID TTL is a duration, as is the response time of making a request to the datastore plugin. The granularity of buckets in these two situations will vastly differ. Trying to pick one set of buckets to rule them all will likely be problematic in the future.

minor nit that if we get in a change to have TTL as a metric, `AddSample` would make more sense than `MeasureSince`. Granted the bucketing for that is also not great for TTLs today but it would be a more intuitive call. It may also make sense to have % categories of TTL remaining or similar instead of precise TTLs, or to be emitting such metrics on the user integration end of things.

> The preferred usage for tally is to provide your own buckets to Scope.Histogram and cache that object locally. We'll need to investigate going forward how to best accomplish this without sacrificing the ability to use other metrics backends.

this somewhat wraps back to the future of spire metrics libraries and design #1040

File: pkg/common/telemetry/m3.go (hunk 3 above)
Comment (amoore877) [reply pkg/common/telemetry/m3.go]: million

**Comment by amoore877** [comment]
just using default buckets

**Comment by evan2645** [comment]
@amoore877 ICYMI, the tests failed with the below error

```
--- FAIL: TestAddSampleForValueHistogram (0.00s)
    m3_test.go:116: 
        	Error Trace:	m3_test.go:116
        	Error:      	Should NOT be empty, but was map[]
        	Test:       	TestAddSampleForValueHistogram
    m3_test.go:117: 
        	Error Trace:	m3_test.go:117
        	Error:      	Should be empty, but was map[0s:0 10ms:0 25ms:0 50ms:0 75ms:0 100ms:0 200ms:0 300ms:0 400ms:0 500ms:0 600ms:0 800ms:0 1s:0 2s:0 5s:0 2562047h47m16.854775807s:1]
        	Test:       	TestAddSampleForValueHistogram
FAIL
FAIL	github.com/spiffe/spire/pkg/common/telemetry	0.157s
```

**Comment by amoore877** [comment]
> @amoore877 ICYMI, the tests failed with the below error
> 
> ```
> --- FAIL: TestAddSampleForValueHistogram (0.00s)
>     m3_test.go:116: 
>         	Error Trace:	m3_test.go:116
>         	Error:      	Should NOT be empty, but was map[]
>         	Test:       	TestAddSampleForValueHistogram
>     m3_test.go:117: 
>         	Error Trace:	m3_test.go:117
>         	Error:      	Should be empty, but was map[0s:0 10ms:0 25ms:0 50ms:0 75ms:0 100ms:0 200ms:0 300ms:0 400ms:0 500ms:0 600ms:0 800ms:0 1s:0 2s:0 5s:0 2562047h47m16.854775807s:1]
>         	Test:       	TestAddSampleForValueHistogram
> FAIL
> FAIL	github.com/spiffe/spire/pkg/common/telemetry	0.157s
> ```

it's always the time you forget to run the tests :) sorry about that

**Comment by amoore877** [comment]
talked offline with @azdagron 
since tally's default buckets don't include proper value buckets, we will have our own for now of an exponential bucket set:
0-1, 1-10, 10-100, 100-1000,.. up to 10^5

**Pull request merged**